base/MAPL_Resource.F90

```diff
@@ -0,0 +1,411 @@
+      type(ESMF_Config), intent(inout) :: config
+      character(len=*), intent(in) :: label
+      character(len=*), intent(in) :: compname
+      logical, intent(out) :: label_is_present
+      character(len=:), allocatable, intent(out) :: label_to_use
+      integer, optional, intent(out) :: rc
+
+      character(len=ESMF_MAXSTR), allocatable :: labels_to_try(:)
+      integer :: i
+      integer :: status
+
+      label_is_present = .false.
+      labels_to_try = get_labels_with_prefix(compname, label)
+
+      do i = 1, size(labels_to_try)
+         label_to_use = trim(labels_to_try(i))
+         call ESMF_ConfigFindLabel(config, label = label_to_use, isPresent = label_is_present, _RC)
+
+         if (label_is_present) then
+            exit
+         end if
+      end do
+
+      _RETURN(_SUCCESS)
+
+   end subroutine get_label_to_use
+
+   ! Find value of scalar variable in config
+   subroutine MAPL_GetResource_config_scalar(config, val, label_to_find, default, compname, rc)
+      type(ESMF_Config), intent(inout) :: config
+      class(*), intent(inout) :: val
+      character(len=*), intent(in) :: label_to_find
+      class(*), optional, intent(in) :: default
+      character(len=*), optional, intent(in) :: compname
+      integer, optional, intent(out) :: rc
+
+      integer :: status, printrc
+      logical :: default_is_present, label_is_present
+      character(len=:), allocatable :: label_to_print
+      character(len=:), allocatable :: label
+
+      default_is_present = present(default)
+
+      if (default_is_present) then
+         _ASSERT(same_type_as(val, default), "Value and default must have same type")
+      end if
+
+      ! If compname is present, find label in some form in config. Else search
+      ! for exact label
+      if (present(compname)) then
+         call get_label_to_use(config, label_to_find, compname, label_is_present, label, _RC)
+      else
+         label = label_to_find
+         call ESMF_ConfigFindLabel(config, label = label, isPresent = label_is_present, _RC)
+      end if
+
+      ! No default and not in config, error
+      ! label or default must be present
+      if (.not. label_is_present .and. .not. default_is_present) then
+         if (present(rc)) rc = ESMF_FAILURE
+         return
+      end if
+
+      select type(val)
+      type is(integer(int32))
+         if (default_is_present .and. .not. label_is_present) then
+            select type(default)
+            type is(integer(int32))
+               val = default
+            end select
+         else
+            call ESMF_ConfigGetAttribute(config, val, label = label, _RC)
+         end if
+      type is(integer(int64))
+         if (default_is_present .and. .not. label_is_present) then
+            select type(default)
+            type is(integer(int64))
+               val = default
+            end select
+         else
+            call ESMF_ConfigGetAttribute(config, val, label = label, _RC)
+         end if
```

Comment:
This might be the occasion where a small include file is preferable.   You would use the FPP preprocessor to define the type and then include the logic  which is otherwise identical in each block.

We can justify it as a first step towards when Fortran supports templates.

Reply:
Actually, we could try it first with a (very long) FPP macro?
Something like:
```
#define TYPE_BLOCK(T) \
     type is (T) \
     if (default_is_present .and. .not. label_is_present) then \
            select type(default) \
            type is(T) \
               val = default \
            class default
               _FAIL("Type of 'default' does not match type of 'val'.")
            end select \
         else \
            call ESMF_ConfigGetAttribute(config, val, label = label, _RC) \
         end if

Reply:
Done.